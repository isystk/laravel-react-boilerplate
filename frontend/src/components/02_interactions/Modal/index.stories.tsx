import { Meta, Story } from '@storybook/react'
import React from 'react'
import Modal, { ModalProps } from './index'
export default {
  title: '02_interactions/Modal',
  component: Modal,
} as Meta

const Template: Story = () => {
  const props: ModalProps = {
    isOpen: true,
  }
  return (
    <Modal {...props}>
      <svg
        aria-hidden="true"
        className="mx-auto mb-4 text-gray-400 w-14 h-14 dark:text-gray-200"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          d="M12 8v4m0 4h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
        ></path>
      </svg>
      <h3 className="mb-5 text-lg font-normal text-gray-500 dark:text-gray-400">
        削除します。よろしいですか？
      </h3>
    </Modal>
  )
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
