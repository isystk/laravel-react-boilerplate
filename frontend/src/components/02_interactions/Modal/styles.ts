import { css, keyframes } from '@emotion/css'

export const modal = css``
