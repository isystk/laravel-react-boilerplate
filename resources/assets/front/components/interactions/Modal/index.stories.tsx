import { JSX, useState } from 'react';
import Modal from './index';
import type { Meta, StoryFn } from '@storybook/react';

export default {
  title: 'Components/Interactions/Modal',
  component: Modal,
  tags: ['autodocs'],
} as Meta<typeof Modal>;

export const Default: StoryFn = () => {
  const [isOpen, setIsOpen] = useState(false);
  return (
    <>
      <button className="btn btn-primary" onClick={() => setIsOpen(true)}>
        Open Modal
      </button>
      <Modal isOpen={isOpen} onClose={() => setIsOpen(false)}>
        <div>モーダルの中身</div>
      </Modal>
    </>
  );
};

export const Small: () => JSX.Element = () => {
  const [isOpen, setIsOpen] = useState(false);
  return (
    <>
      <button className="btn btn-primary" onClick={() => setIsOpen(true)}>
        Open Small Modal
      </button>
      <Modal isOpen={isOpen} onClose={() => setIsOpen(false)} small>
        <div>小さなモーダル</div>
      </Modal>
    </>
  );
};
