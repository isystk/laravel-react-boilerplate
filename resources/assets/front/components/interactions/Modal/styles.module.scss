.overlay {
    position: fixed;
    inset: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    background-color: rgba(0, 0, 0, 0.5);
    z-index: 50;
}

.modal {
    background-color: #ffffff;
    border-radius: 0.5rem;
    box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),
    0 4px 6px -4px rgba(0, 0, 0, 0.1);
    position: relative;
    width: 100%;
    max-width: 50rem;
    margin: 0 1rem;
    &.small {
        max-width: 30rem;
    }
}

.closeButton {
    position: absolute;
    top: 0.75rem; // 12px
    right: 0.75rem; // 12px
    color: #6b7280; // Tailwind's gray-500
    background: none;
    border: none;
    cursor: pointer;
    font-size: 1.25rem;

    &:hover {
        color: #374151; // Tailwind's gray-700
    }
}

.content {
    padding: 1.5rem; // 24px
}
