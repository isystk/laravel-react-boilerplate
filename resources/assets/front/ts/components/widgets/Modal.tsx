import React, { FC } from "react";
import Portal from "@/components/commons/Portal";

type Props = {
    children: React.ReactNode;
    isOpen: boolean;
    handleClose: () => void;
};

const Modal: FC<Props> = ({ children, isOpen, handleClose }) => {
    return (
        <Portal>
            {isOpen && <div id="overlay-background"></div>}
            <div className={`isystk-overlay ${isOpen ? "open" : ""}`}>
                <button
                    type="button"
                    className="close"
                    aria-label="Close"
                    onClick={handleClose}
                >
                    <span aria-hidden="true">&times;</span>
                </button>
                {children}
            </div>
        </Portal>
    );
};

export default Modal;
