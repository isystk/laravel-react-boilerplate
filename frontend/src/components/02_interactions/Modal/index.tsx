import React, { FC, useEffect, useRef, useState } from 'react'
import ReactDOM from 'react-dom'
import { ContainerProps, WithChildren } from 'types'
import * as styles from './styles'
import { connect } from '@/components/hoc'

/** ModalProps Props */
export type ModalProps = WithChildren & {
  isOpen: boolean
  acceptLabel?: string
  cancelLabel?: string
  handleAccept?: () => void
  handleCancel?: () => void
}
/** Presenter Props */
export type PresenterProps = ModalProps & { onClose; Portal }

/** Presenter Component */
const ModalPresenter: FC<PresenterProps> = ({
  children,
  isOpen,
  acceptLabel = 'はい',
  cancelLabel = 'いいえ',
  handleAccept = () => ({}),
  handleCancel = () => ({}),
  Portal,
  ...props
}) => (
  <>
    <Portal>
      {isOpen && (
        <>
          <div className="bg-gray-900 bg-opacity-50 dark:bg-opacity-80 fixed inset-0 z-40"></div>
          <div
            id="popup-modal"
            className={`${
              isOpen ? '' : 'hidden '
            } fixed top-0 left-0 right-0 z-50 p-4 overflow-x-hidden overflow-y-auto md:inset-0 h-[calc(100%-1rem)] max-h-full`}
          >
            <div className="relative w-full max-w-md max-h-full m-auto">
              <div className="relative bg-base rounded-lg shadow dark:bg-gray-700">
                <button
                  type="button"
                  className="absolute top-3 right-2.5 text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm p-1.5 ml-auto inline-flex items-center dark:hover:bg-gray-800 dark:hover:text-white"
                  data-modal-hide="popup-modal"
                  onClick={() => handleCancel()}
                >
                  <svg
                    aria-hidden="true"
                    className="w-5 h-5"
                    fill="currentColor"
                    viewBox="0 0 20 20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      fillRule="evenodd"
                      d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
                      clipRule="evenodd"
                    ></path>
                  </svg>
                  <span className="sr-only">Close modal</span>
                </button>
                <div className="p-6 text-center">
                  {children}
                  <button
                    data-modal-hide="popup-modal"
                    type="button"
                    className="text-white bg-red-600 hover:bg-red-800 focus:ring-4 focus:outline-none focus:ring-red-300 dark:focus:ring-red-800 font-medium rounded-lg text-sm inline-flex items-center px-5 py-2.5 text-center mr-2"
                    onClick={() => handleAccept()}
                  >
                    {acceptLabel}
                  </button>
                  <button
                    data-modal-hide="popup-modal"
                    type="button"
                    className="text-gray-500 bg-base hover:bg-main focus:ring-4 focus:outline-none focus:ring-gray-200 rounded-lg border border-gray-200 text-sm font-medium px-5 py-2.5 hover:text-gray-900 focus:z-10 dark:bg-gray-700 dark:text-gray-300 dark:border-gray-500 dark:hover:text-white dark:hover:bg-gray-600 dark:focus:ring-gray-600"
                    onClick={() => handleCancel()}
                  >
                    {cancelLabel}
                  </button>
                </div>
              </div>
            </div>
          </div>
        </>
      )}
    </Portal>
  </>
)

/** Container Component */
const ModalContainer: React.FC<ContainerProps<ModalProps, PresenterProps>> = ({
  presenter,
  children,
  ...props
}) => {
  const Portal: FC = ({ children }) => {
    const ref = useRef<HTMLElement>()
    const [mounted, setMounted] = useState(false)

    useEffect(() => {
      const current = document.querySelector<HTMLElement>('body')
      if (current) {
        ref.current = current
      }
      setMounted(true)
    }, [])

    return mounted
      ? ReactDOM.createPortal(
          <>{children}</>,
          ref.current ? ref.current : new Element()
        )
      : null
  }
  return presenter({ children, Portal, ...props })
}

export default connect<ModalProps, PresenterProps>(
  'Modal',
  ModalPresenter,
  ModalContainer
)
