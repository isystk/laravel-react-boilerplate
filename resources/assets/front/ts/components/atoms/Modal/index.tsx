import React from "react";
import styles from './styles.module.scss';
import Portal from "@/components/atoms/Portal";

type Props = {
    isOpen: boolean;
    onClose: () => void;
    children: React.ReactNode;
}

const Modal = ({ isOpen, onClose, children }: Props) => {
    if (!isOpen) return null;

    return (
        <Portal>
            <div className={styles.overlay} onClick={onClose}>
                <div className={styles.modal} onClick={(e) => e.stopPropagation()}>
                    <button
                        type="button"
                        aria-label="Close"
                        onClick={onClose}
                        className={styles.closeButton}
                    >
                        <span aria-hidden="true">Ã—</span>
                    </button>
                    <div className={styles.content}>
                        {children}
                    </div>
                </div>
            </div>
        </Portal>
    );
};

export default Modal;
