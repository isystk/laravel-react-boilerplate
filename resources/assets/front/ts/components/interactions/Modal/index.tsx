import styles from './styles.module.scss';
import Portal from '@/components/interactions/Portal';
import { JSX } from 'react';

type Props = {
  isOpen: boolean;
  onClose: () => void;
  children: JSX.Element;
};

const Modal = ({ isOpen, onClose, children }: Props) => {
  if (!isOpen) return null;

  return (
    <Portal>
      <div className={styles.overlay} onClick={onClose}>
        <div className={styles.modal} onClick={e => e.stopPropagation()}>
          <button type="button" aria-label="Close" onClick={onClose} className={styles.closeButton}>
            <span aria-hidden="true">Ã—</span>
          </button>
          <div className={styles.content}>{children}</div>
        </div>
      </div>
    </Portal>
  );
};

export default Modal;
