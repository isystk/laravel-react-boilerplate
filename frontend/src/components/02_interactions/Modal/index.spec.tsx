import React from 'react'
import renderer from 'react-test-renderer'
import Modal, { ModalProps } from './index'

describe('Modal', () => {
  it('Match Snapshot', () => {
    const props: ModalProps = {
      isOpen: true,
    }
    const component = renderer.create(<Modal {...props} />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
