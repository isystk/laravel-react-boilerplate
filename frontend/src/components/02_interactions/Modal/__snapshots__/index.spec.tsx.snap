// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Match Snapshot 1`] = `null`;
